File: SortVision/public/code/heap/golang/heapSort.go
package main

import (
	"fmt"
)

// heapify ensures the heap property is maintained in a binary tree.
// Parameters:
// - arr: The slice of integers representing the heap.
// - n: The size of the heap.
// - i: The index of the current node to heapify.
func heapify(arr []int, n int, i int) {
    largest := i
    left := 2*i + 1
    right := 2*i + 2

    // Check if the left child exists and is greater than the root
    if left < n && arr[left] > arr[largest] {
        largest = left
    }

    // Check if the right child exists and is greater than the largest so far
    if right < n && arr[right] > arr[largest] {
        largest = right
    }

    // If the largest is not the root, swap and recursively heapify the affected subtree
    if largest != i {
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)
    }
}

// HeapSort sorts an array of integers using the heap sort algorithm.
// Parameters:
// - arr: The slice of integers to be sorted.
// Returns:
// - A new slice containing the sorted elements.
func HeapSort(arr []int) []int {
    n := len(arr)
    if n <= 1 {
        // Return a copy of the original array if it has one or no elements
        return append([]int{}, arr...)
    }

    // Copy the array to avoid modifying the original
    nums := append([]int{}, arr...)

    // Build the heap (rearrange the array)
    for i := n/2 - 1; i >= 0; i-- {
        heapify(nums, n, i)
    }

    // Extract elements from the heap one by one
    for i := n - 1; i > 0; i-- {
        // Move the current root (largest) to the end
        nums[0], nums[i] = nums[i], nums[0]
        // Call heapify on the reduced heap
        heapify(nums, i, 0)
    }

    return nums
}

// Example usage of the HeapSort function.
func main() {
    fmt.Println("Heap Sort Implementation in Go")

    // Example arrays
    testCases := [][]int{
        {64, 34, 25, 12, 22, 11, 90},
        {},
        {42},
        {1, 2, 3, 4, 5},
        {5, 4, 3, 2, 1},
        {3, 1, 3, 1, 3},
        {3, 1, 4, 1, 5, 9, 2, 6},
        {100, 50, 25, 12},
        {-5, -2, -8, -1},
    }

    for _, arr := range testCases {
        sorted := HeapSort(arr)
        fmt.Printf("Original: %v, Sorted: %v\n", arr, sorted)
    }
}

File: SortVision/public/code/heap/csharp/heapSort.cs
using System;

public class HeapSort
{
    // Main sorting method
    public static void Sort(int[] arr)
    {
        // Validate input
        if (arr == null)
            throw new ArgumentNullException(nameof(arr), "Input array cannot be null");
        
        int n = arr.Length;
        if (n <= 1) return;  // Already sorted for empty/single-element arrays

        // Build initial max-heap
        BuildMaxHeap(arr);
        
        // Extract elements one by one from heap
        for (int i = n - 1; i > 0; i--)
        {
            // Move current root to end
            Swap(arr, 0, i);
            
            // Heapify the reduced heap
            Heapify(arr, i, 0);
        }
    }

    // Maintain max-heap property for a subtree
    private static void Heapify(int[] arr, int heapSize, int index)
    {
        int largest = index;            // Initialize largest as root
        int leftChild = 2 * index + 1;  // Left child index
        int rightChild = 2 * index + 2; // Right child index

        // If left child is larger than root
        if (leftChild < heapSize && arr[leftChild] > arr[largest])
            largest = leftChild;
        
        // If right child is larger than current largest
        if (rightChild < heapSize && arr[rightChild] > arr[largest])
            largest = rightChild;
        
        // If largest is not root
        if (largest != index)
        {
            Swap(arr, index, largest);
            
            // Recursively heapify the affected subtree
            Heapify(arr, heapSize, largest);
        }
    }

    // Build max-heap from unordered array
    private static void BuildMaxHeap(int[] arr)
    {
        int n = arr.Length;
        
        // Start from last non-leaf node and work backwards
        for (int i = n / 2 - 1; i >= 0; i--)
        {
            Heapify(arr, n, i);
        }
    }

    // Swap array elements
    private static void Swap(int[] arr, int i, int j)
    {
        // Skip swap if indices are identical
        if (i == j) return;
        
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

class Program
{
    static void Main()
    {
        // Example usage
        int[] arr = { 12, 11, 13, 5, 6, 7 };
        Console.WriteLine("Original array:");
        PrintArray(arr);

        HeapSort.Sort(arr);
        
        Console.WriteLine("\nSorted array:");
        PrintArray(arr);
        
        // Execute test cases
        RunTestCases();
    }

    static void PrintArray(int[] arr)
    {
        Console.WriteLine(string.Join(" ", arr));
    }

    static void RunTestCases()
    {
        Console.WriteLine("\nRunning test cases...");
        
        TestCase(new int[] { }, "Empty array");
        TestCase(new int[] {1}, "Single element");
        TestCase(new int[] {5, 1, 4, 2, 8}, "Unsorted array");
        TestCase(new int[] {1, 2, 3, 4, 5}, "Already sorted");
        TestCase(new int[] {5, 4, 3, 2, 1}, "Reverse sorted");
        TestCase(new int[] {3, 1, 4, 1, 5, 9, 2, 6}, "Array with duplicates");
        TestCase(new int[] {-5, -1, -3, 2, 0}, "Array with negatives");
    }

    static void TestCase(int[] arr, string description)
    {
        Console.Write($"\n{description}: ");
        PrintArray(arr);
        
        try
        {
            int[] original = (int[])arr.Clone();
            HeapSort.Sort(arr);
            
            if (IsSorted(arr))
                Console.WriteLine("PASS");
            else
                Console.WriteLine($"FAIL - Sorted: {string.Join(" ", arr)}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"ERROR: {ex.Message}");
        }
    }

    static bool IsSorted(int[] arr)
    {
        for (int i = 1; i < arr.Length; i++)
            if (arr[i] < arr[i - 1])
                return false;
        return true;
    }
}